오픈마켓 [STEP1] 아샌, 제인

ICS-Asan opened the pull request:
안녕하세요. 코리! @corykim0829
아샌(@ICS-Asan), 제인(@siwonkim0) 입니다. 

이번 프로젝트 첫 PR 올립니다! 잘 부탁드려요!! 

감사합니다!!

## 💻    주요 구현 내용

### < ProductList >

- 서버 API 문서를 참고하여 모델 타입 구현
- 상품의 images, vendors프로퍼티의 타입을 옵셔널로 설정
    - 상품 리스트를 조회할 때는 images, vedors프로퍼티가 사용되지 않음
    - 기존에는 상품 리스트를 조회할 때 사용되는 타입, 상품 상세를 조회할 때 사용되는 타입으로 각각 구현
    - 동일한 코드가 반복되는 것을 줄이기 위해 두 타입을 합친 후 images, vendors에 옵셔널 타입으로 설정하여 두가지 상황에서 모두 사용할 수 있도록 구현

### < JSONParser >

- JSON데이터를 디코딩하고 에러를 처리하는 decodeData메서드 구현
    - JSON데이터를 디코딩 하기 위한 decode 메서드는 기본적으로 에러를 throw하는 메서드
    - 외부에서 decode 할 때마다 에러를 처리하는 것 보다 decode와 동시에 에러를 처리하게 하여 호출부에서 코드를 간결하게 할 수 있음
    - 제네릭 타입을 적용하여 메서드 호출시에 원하는 타입으로 디코딩 할 수 있도록 구현

### < APIManager >

- 네트워크 통신을 통해 받은 데이터를 담아둘 수 있는 프로퍼티 구현(`apiHealth`, `product`, `productList`)
    - dataTask메서드에서 값을 리턴하지 않고 외부에 프로퍼티를 생성해두고 값을 넣어주는 방식으로 구현
    - 네트워크 통신의 결과를 프로퍼티를 편리하게 사용할 수 있도록 하기 위해 생성
- 프로퍼티가 업데이트 된 이후 사용되도록 하기 위해 DataTask가 사용되는 곳에 Semaphore를 사용하여 제어

### < WorkType >

- enum 타입으로 URLRequest에서 사용되는 URL을 한번에 관리
- 작업 종류에 따라 사용되는 URL을 연산 프로퍼티로 정의
- URLComponents 사용
    - 쿼리가 사용된 URL을 생성하는데 필요한 String을 직접 작성하는 것은 가독성도 좋지 않고 관리도 어렵다고 판단
    - name-value의 쌍을 URLQueryItem 이니셜라이저를 통해 생성
    - 어떤 QueryItem이 어떤 name과 value를 통해 생성되는지 파악하기가 편해짐

```swift
enum WorkType {
    static let apiHost: String = "https://market-training.yagom-academy.kr/"
    case checkProductList(pageNumber: Int, itemsPerPage: Int)
    
    var url: URL? {
        switch self {
        case .checkProductList(let pageNumber, let itemsPerPage):
            var components = URLComponents(string: WorkType.apiHost + "api/products")
            let number = URLQueryItem(name: "page_no", value: "\(pageNumber)")
            let items = URLQueryItem(name: "items_per_page", value: "\(itemsPerPage)")
            components?.queryItems = [number, items]
            return components?.url
        }
    }
}
```

### < URLRequest >

- URLRequest 를 extension하여 새로운 이니셜라이저를 정의
- 파라미터로 url과 httpMethod를 받아서 URLRequest가 쓰이는 곳에서 코드가 간결해질 수 있도록 구현

## ❓   질문할 내용

### < 비동기적으로 실행되는 클로저와 메인에서의 UI 업데이트 >

dataTask 메서드 안에서 `main.async` 로 뷰를 업데이트하면 정상적으로 작동하는것을 확인했습니다. 

뷰를 바로 업데이트하기 위해서는, dataTask 메서드를 뷰컨트롤러에서 구현해야하는데.. 뷰컨이 너무 비대해져서 모델에 구현하려고 하다 보니 문제가 생겼습니다.

dataTask내부의 코드블럭은 비동기적으로 작동하기 때문에 내부에서 프로퍼티를 업데이트 하는 작업 또한 비동기적으로 진행됩니다.

그래서 해당 프로퍼티를 이용해서 뷰컨에서 작업을 하려고 하면 비동기적으로 진행되는 작업이 끝나지 않은 상태에서 프로퍼티에 접근을 하게됩니다.

그로 인해  프로퍼티가 업데이트 되지 않고 사용되는 상황이 발생하게 되었습니다.

문제해결을 위해.. apiStatus 프로퍼티를 업데이트한 이후에 사용하고자 `DispatchSemaphore`를 이용하여 메인스레드가 작업을 기다리도록 구현하게 되었습니다. 

현재는 개선할 방법이 semaphore를 이용하는 방법 외에는 떠오르지 않아서 이렇게 구현했지만...

업데이트해야할 UI요소가 많아지게되면 이전 작업이 끝날때 까지 기다리는 것을 반복하며 문제가 생길것 같고 

또 메인스레드를 기다리게 하는 방법은 좋지 않다고 들어서 다른 방향으로 해결할 수 있는 방법이 있는지 궁금합니다. 

![image](https://user-images.githubusercontent.com/91048487/148352756-13b140e0-cabd-4384-9c39-0b25acbad48f.png)
```swift
class APIManager {
    
    let apiHost: String = "https://market-training.yagom-academy.kr/"
    var apiStatus: String?
    let semaphore = DispatchSemaphore(value: 0)
    
    func healthChecker() {
        let url: String = apiHost + "healthChecker"
        var request = URLRequest(url: URL(string: url)!,timeoutInterval: Double.infinity)
        request.httpMethod = "GET"
        
        let task = URLSession.shared.dataTask(with: request) { data, response, error in
            guard let data = data else {
                print(String(describing: error))
								self.semaphore.signal()
                return
            }
            self.apiStatus = String(data: data, encoding: .utf8)!
	          **self.semaphore.signal()**
        }
        task.resume()
        semaphore.wait()
    }
    
}

class ViewController: UIViewController {

    let api = APIManager()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        api.checkAPIHealth()
        print(api.apiHealth)
    }
}
```
감사합니다!!

ICS-Asan committed:
feat: Mock데이터 파일 에셋에 추가 및 Model폴더 생성 #1

ICS-Asan committed:
feat: OpenMarket 파일 생성 #1

siwonkim0 committed:
feat: OpenMarket, OpenMarketPage 타입 #1

ICS-Asan committed:
feat: JSONParser 타입 구현 #1

- 내부 decodeData 타입 메서드 구현
- Data 타입의 JSON데이터를 디코딩 할 수 있도록 구현

ICS-Asan committed:
file: gitignore pods 주석 해제

- cocoapods을 업로드 하지 않도록 gitignore적용

siwonkim0 committed:
file: cocoapods 적용

siwonkim0 committed:
file: cocoapods 적용

siwonkim0 committed:
Merge branch 'step1' of https://github.com/ICS-Asan/ios-open-market into step1

ICS-Asan committed:
test: OpenMarketTests 생성 및 TestCase 구현 #2

siwonkim0 committed:
feat: APIManager 타입 구현 #3

- get메서드로 API를 가져오는 healthChecker 메서드 구현

ICS-Asan committed:
feat: checkProductDetail 메서드 구현 #3

- get메서드로 API를 가져오는 checkProductDetail 메서드 구현

siwonkim0 committed:
test: APIManagerTests 생성 및 TestCase 구현 #4

ICS-Asan committed:
feat: checkProductList 메서드 구현 #3 #4

- productList 프로퍼티 생성
- 테스트 케이스를 통해 정상 작동 확인

siwonkim0 committed:
refactor: JSONParser 타입의 decodeData 메서드에서 에러처리 #1

ICS-Asan committed:
feat: URLManager 타입 구현 #1

siwonkim0 committed:
feat: URLRequest extension 구현 #3

- 처리할 업무타입에 따라 URLRequest 객체를 생성하는 이니셜라이저 구현

ICS-Asan committed:
refactor: JSON모델 타입 미구현부 추가 #1

- images, vendors 누락확인 후 추가 구현

siwonkim0 committed:
refactor: JSON 모델 타입 수정 #1

- ProductList, ProductDetail 네이밍 수정
- ProductSummary 타입 구현

ICS-Asan committed:
refactor: ProductDetail 내부 프로퍼티 옵셔널타입 설정 #1

- ProductSummary에 포함되지 않는 images,vendors프로퍼티를 추가로 받는 타입 생성
- 불필요한 반복코드를 작성하지 않게 하기 위해 두 타입을 하나로 병합
- 추가되는 프로퍼티에 옵셔널 타입으로 설정하여 정상적으로 디코딩되도록 함
- 타입의 재사용성을 높임

siwonkim0 committed:
refactor: URL 옵셔널 바인딩 처리 #3

ICS-Asan committed:
refactor: WorkType url 연산 프로퍼티 URLQueryItem 적용 #3

- checkProductDetail의 연관값을 이용하여 url 생성

siwonkim0 committed:
refactor: timeInterval 설정부분 삭제 #3

corykim0829 commented on OpenMarket/OpenMarket/Model/WorkType.swift:
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+enum WorkType {
```
여기 enum의 네이밍이 왜 WorkType인지 이유가 궁금해요! 👀

siwonkim0 replied:
작업의 종류를 담은 타입이고, 작업의 속성이 URL이라고 생각하여 WorkType로 네이밍했는데 생각해보니 URL을 담고있는 타입이기때문에 URLManager로 수정하는게 좋을 것 같습니다. 😄

corykim0829 commented on OpenMarket/OpenMarket/Model/ProductList.swift:
```diff
@@ -0,0 +1,68 @@
+    let lastPage: Int
+    let hasNext, hasPrev: Bool
+
+    enum CodingKeys: String, CodingKey {
```
CodingKey를 잘 사용하셨네요!
이와 같이 스네이크 케이스를 카멜케이스로만 바꾸는 케이스만 있을 때에는  `keyDecodingStrategy`, `keyEncodingStrategy`를 사용하면 조금 더 간단하게 원하는 프로퍼티명으로 디코딩, 인코딩할 수 있어요!
참고만 하셔도됩니다!

corykim0829 commented on OpenMarket/OpenMarket/Model/ProductList.swift:
```diff
@@ -0,0 +1,68 @@
+    }
+}
+
+struct ProductDetail: Codable {
```
하나의 객체에 되도록 하나의 파일로 처리하는 게 유지보수에 좋습니다!

siwonkim0 replied:
피드백 반영해서 파일분리했습니다~! 😁

corykim0829 commented on OpenMarket/OpenMarket/Model/WorkType.swift:
```diff
@@ -0,0 +1,24 @@
+    static let apiHost: String = "https://market-training.yagom-academy.kr/"
+    case healthChecker
+    case checkProductDetail(id: Int)
+    case checkProductList(pageNumber: Int, itemsPerPage: Int)
```
enum case로 URL처리를 잘 했네요 👍

siwonkim0 replied:
😁

corykim0829 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,58 @@
+import Foundation
+
+class APIManager {
```
APIManager를 통해서 네트워크 통신을 관리하는 것은 매우 좋은 접근이라고 생각합니다.
하지만 이렇게 APIManager 안에 특정 네트워크 통신 값을 지정해두는 것은 해당 값이 들어올 시점을 한번 더 계산해줘야하는 일, 그리고 옵셔널 처리를 한번 더 해줘야하는 번거로움이 생기게 됩니다. 
각 함수에서 통신 결과값을 반환해보도록 처리해보는 방법을 고민해보면 좋을 것 같아요!

siwonkim0 replied:
저희도 피드백 주신 부분에 대해 많은 고민이 있었습니다...
값을 리턴하는 방법에 대해 생각해보다가 리턴이 어려워서 프로퍼티에 할당하는 식으로 구현했었는데, 좀 더 알아보니 completion handler을 사용해서 Result 타입으로 데이터를 반환하는 방법을 알게되어 그렇게 수정하였습니다!!

corykim0829 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,58 @@
+        semaphore.wait()
+    }
+    
+    func checkProductDetail(from id: Int) {
```
여기에서는 id 파라미터 하나로 쓰는게 외부에서 이 함수를 호출하여 사용할때에도 더 명확한 것 같아요. from만 있으면 어떤 값을 넣어야할지 애매한 것 같아요!

ICS-Asan replied:
피드백 주신 부분에 대해 생각해보니 checkProductDetail를 호출하는 부분에서 어떤 값을 넣어야할지 애매할 것 같다는 피드백에 공감이 되었습니다.
그래서 checkProductList메서드와 동일하게 어떤 값을 받을지 명확하게 하고자 from을 삭제하고 id만 사용하도록 수정하였습니다!

corykim0829 commented on OpenMarket/OpenMarket/Model/JSONParser.swift:
```diff
@@ -0,0 +1,17 @@
+
+enum JSONParser {
+    
+    static func decodeData<T: Decodable>(of JSON: Data, how: T.Type) -> T? {
```
how보다는 type과 같이 명사형을 사용하는게 더 명확할 것 같아요! 
how라고 하면 뭔가 동작이 들어가야할 것 같은 느낌이 들어서요 :)

ICS-Asan replied:
how->type로 네이밍을 수정하였습니다!!
감사합니다!!

corykim0829 reviewed (changes requested):
수고 정말 많으셨습니다! 아샌, 제인 🙌
두분이서 많이 고민한 흔적이 느껴지는 코드들이였습니다! 👏👏👏

---

### 질문할 내용

> 비동기적으로 실행되는 클로저와 메인에서의 UI 업데이트

먼저 뷰컨트롤러가 비대해진다고 해서 모델에 구현하려고 하는 과정에서 문제가 생겼다고 하셨는데, 뷰컨트롤러에 모두 구현한다고해서 꼭 나쁘다고는 할 수 없습니다. 그래서 먼저 뷰컨트롤러에서 잘 작동하는 코드를 구현 후에 분리시켜보는 것도 좋은 접근일 것 같아요 :)

비동기로 처리되는 API에 대해서는 제가 리뷰에서 조금 이야기를 하긴 했는데요! 현재 두분이 걱정하는 부분은 세마포어를 사용하지 않고 다시 한번 코드를 작성해보면 좋을 것 같습니다!

세말씀하신대로 세마포어는 사용성을 매우 저하시키기 때문에 지양해야합니다 🥲

답을 제가 바로 알려드리기 보다는 비동기 처리에 대해서 조금 더 고민해보고 찾아보시고 이야기해보면 좋을 것 같습니다!!

---

﹢추가적으로 먼저 mock json 파일로 스텝을 먼저 진행하는 걸로 알고 있었는데, 이 부분은 바로 서버로 하라고 하셨나요?
여유가 된다면 mock json 파일도 파싱해보는 것도 해보시고
이번에 하신 방법대로 서버 통신으로 구현하신 부분은 네트워크 요청 부분을 로컬에서 테스트하는 방법도 한번 찾아보고 도전해보는 것도 좋을 것 같습니다 😀 아래 영상을 참고하면 아주 좋습니다 👍
https://www.youtube.com/watch?v=meTnd09Pf_M

ICS-Asan committed:
refactor: dataTask호출부 creatDataTask메서드로 분리 #3

- 반복되는 dataTask호출을 메서드로 분리하여 사용
- 제네릭 타입으로 구현하여 재사용성을 높임

siwonkim0 committed:
file: 파일 분리 #5

ICS-Asan committed:
refactor: WorkType -> URLManager 네이밍 변경 #5

siwonkim0 committed:
refactor: checkProductDetail, JSONParser 파라미터 수정 #5

- 관련 test 코드도 수정

corykim0829 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -0,0 +1,61 @@
+        let task = URLSession.shared.dataTask(with: request) { data, response, error in
+            guard let data = data else {
+                print(String(describing: error))
+                completion(.failure(error!))
```
강제 언래핑을 사용하지 않고 처리하면 좋을 것 같아요 :)

siwonkim0 replied:
앗 그렇군요! 옵셔널 바인딩처리하였습니다 ~~!

corykim0829 reviewed (changes requested):
수고하셨습니다 👏

ICS-Asan committed:
feat: URLSessionProtocol 생성 및 구현 #6

- URLSession에 URLSessionProtocol을 채택
- APIManager내부에 URLSession 프로퍼티와 이니셜라이저 생성

siwonkim0 committed:
feat: MockURLSession 생성 및 구현 #6

- MockData 생성
- MockURLSessionDataTask 생성

ICS-Asan committed:
file: TestDouble 폴더 생성 및 파일 분리 #6

- 파일 target 설정 변경

siwonkim0 committed:
test: APIManagerTests 재구현 #6

ICS-Asan committed:
refactor: checkAPIHealth 내부 옵셔널 처리, 에러처리 수정 #6

- response가 없는 error를 처리하기 위해 에러처리 로직 순서 변경

ICS-Asan commented:
1.  세마포어를 사용하지 않기 위해 escaping comletion closure을 활용하여 성공과 실패시 저희가 원하는 값을 넘겨줄 수 있도록 구현하였습니다. 추후에 뷰컨에서 값을 받아 바로 사용할 수 있게 될 것 같습니다.
2. MockURLSession을 구현하여 네트워킹 없이 테스트할 수 있도록 하였습니다. 현재 저희가 구현한 코드와 파일 분리에 대해 살펴봐주시고 개선할 부분이 있다면 피드백 주시면 감사하겠습니다.😃
3. 말씀해주신 `keyDecodingStrategy`을 사용하는 것에 대해서 저희도 구현 당시에 고민이 많았습니다. 고민 결과 `convertFromSnakeCase`를 사용하면 간편하게 카멜케이스로 적용할 수 있지만 네이밍을 서버에 의존적이지 않게 하기 위해 codingkey를 선택하였습니다. 현재는 서버의 네이밍과 같은 상황이지만 추후에 프로퍼티명을 변경할 수도 있다고 판단했습니다.

corykim0829 commented on OpenMarket/OpenMarketTests/TestDouble/MockURLSession.swift:
```diff
@@ -0,0 +1,39 @@
+class MockURLSession: URLSessionProtocol {
+    
+    let sessionDataTask = MockURLSessionDataTask()
+    var makeRequestFail = false
```
Bool 타입 변수를 동사로 시작할 때에는 보통 조동사로 시작해요! 참고하세요 :)

corykim0829 reviewed (approved):
수고 정말 많으셨습니다!! 
네트워킹 타입 테스트 너무 깔끔하게 잘 구현하셨네요!! 👍👍👍

파일 분리도 깔끔하게 하셨습니다! 유지 보수하기 좋은 코드가 되었습니다 :)

테스트 함수명들 또한 명확해서 좋았습니다 😆

다음 스텝 진행해주시고 저에게 목표 일자 말해주세요!
